rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+
+We should introduce a single "feature flag" on the API as an undocumented and
+unstable query parameter `unstable__include_sensitive_results`.
+
+## API versioning
+
+There are no API versioning concerns. We already filter sensitive content based
+on provider data, we'll just be expanding upon that existing feature by
+incorporating a new signal into the filter. I do not believe this requires a new
+API version prefix or distinctly new parameter.
+
+## Areas touched
+
+Implementing this plan will require making changes in the following areas:
+
+- API
+- Elasticsearch
+- Ingestion server
+- Airflow data refresh DAG factory
+
+## Terms
+
+- "Origin index": This is the index we have now that includes all documents.
+- "Filtered index": This is the index we will have after this implementation
+  plan is implemented. Its documents are derived from the origin index but
+  exclude any that have sensitive terms in their textual content.
+
+## Overview of implementation process
+
+In depth descriptions of each of these steps are covered below. Each step will
+be implemented as a single pull request including appropriate tests.
+
+1. Update the ingestion server with a new action to use the reindex API to
+   generate a new index based on the origin index that only includes documents
+   that do not have sensitive terms in the three textual content fields
+   (`title`, `description`, `tags.name`). This is the filtered index.
+   - Requires the following:
+     - SENSITIVE_TERMS injected as new environment variable. See
+       [the related IP](https://github.com/WordPress/openverse/pull/911) for how
+       this will be implemented in Terraform.
+     - Ingestion server action to create the new index using `reindex` and an
+       update the promote action to allow it to promote (point alias) for the
+       filtered index
+     - Updates to the `load_sample_data.sh` script to generate the filtered
+       index
+     - A benign list of terms to use for local testing that matches results in
+       the sample data
+1. Update the data refresh DAG factory code to call the new filtered index
+   actions.
+1. Update the API with the new query parameter and query the filtered index when
+   not including sensitive results.
+1. Update the API to query the filtered index by document `_id` to deduce which
+   documents in a query that includes sensitive results do not have sensitive
+   terms. Use the resulting list of "safe" IDs to inverse the check and mark
+   documents that are sensitive. Simultaneously, update the media serialisers to
+   add the new `sensitivity` field. The serialiser field derives the array
+   entries based on the results of the previous step and the existing `mature`
+   field.
+
+Many thanks to @robfelty for helping me find a much simpler approach to
+sensitive result designation than I had originally devised. This solution is
+easily 10x more intelligible and simpler to document and implement than the
+others I had considered.
```

Comment:
Agreed, this makes a lot more sense on initial read to my Elasticsearch-novice experience!